azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java

```diff
@@ -565,40 +566,74 @@ private void fetchJobTrackerToken(final String userToProxy, final Props props,
-      // check if we get the correct FS, and most importantly, the conf
-      logger.info("Getting DFS token from " + fs.getUri());
-      final Token<?>[] fsTokens =
-          fs.addDelegationTokens(getMRTokenRenewerInternal(new JobConf()).toString(), cred);
-      if (fsTokens.length == 0) {
-        logger.error("Failed to fetch DFS token for ");
-        throw new HadoopSecurityManagerException(
-            "Failed to fetch DFS token for " + userToProxy);
-      }
+      final String renewer = getMRTokenRenewerInternal(new JobConf()).toString();
 
-      for (final Token<?> fsToken : fsTokens) {
-        logger.info(String.format(
-            "DFS token from namenode pre-fetched, token kind: %s, token service: %s",
-            fsToken.getKind(), fsToken.getService()));
-      }
+      // Get the tokens name node
+      fetchNameNodeTokenInternal(renewer, cred, userToProxy, null);
 
       // getting additional name nodes tokens
       final String otherNamenodes = props.get(
           HadoopSecurityManager_H_2_0.OTHER_NAMENODES_TO_GET_TOKEN);
       if ((otherNamenodes != null) && (otherNamenodes.length() > 0)) {
         logger.info("Fetching token(s) for other namenode(s): " + otherNamenodes);
         final String[] nameNodeArr = otherNamenodes.split(",");
-        final Path[] ps = new Path[nameNodeArr.length];
-        for (int i = 0; i < ps.length; i++) {
-          ps[i] = new Path(nameNodeArr[i].trim());
+        for (String nameNode : nameNodeArr) {
+          fetchNameNodeTokenInternal(renewer, cred, userToProxy, new Path(nameNode.trim()).toUri());
         }
-        TokenCache.obtainTokensForNamenodes(cred, ps, HadoopSecurityManager_H_2_0.this.conf);
-        logger.info("Successfully fetched tokens for: " + otherNamenodes);
+      }
+      logger.info("Successfully fetched tokens for: " + otherNamenodes);
+    } else {
+      logger.info(
+              HadoopSecurityManager_H_2_0.OTHER_NAMENODES_TO_GET_TOKEN + " was not configured");
+    }
+  }
+
+  /**
+   * fetchNameNodeInternal - With modified UGI which is of the format,
+   * <userToProxy>/az_<host name>_<exec_id>
+   * Due to this change, the FileSystem cache creates an entry per execution instead of an entry
+   * per proxy user. This could blow up the cache very quickly on a busy Executor and cause OOM.
+   * To make this worse, the entry in Cache is never used as it is specfic to an execution.
+   * To avoid this, always create a new instance of FileSystem locally and close it.
+   *
+   * @param renewer
+   * @param cred
+   * @param userToProxy
+   * @param uri
+   * @throws IOException
+   * @throws HadoopSecurityManagerException
+   */
+
+  private void fetchNameNodeTokenInternal(final String renewer, final Credentials cred,
+          final String userToProxy, final URI uri)
+          throws IOException, HadoopSecurityManagerException {
+    FileSystem fs = null;
+    try {
+      if (uri == null) {
+        fs = FileSystem.get(conf);
```

Comment:
nit: Can you put a comment here explaining that `get(conf)` is important as opposed to `newInstance(...)` because the disable cache flag is only checked on the former?